semantic-core/src/Analysis/FlowInsensitive.hs

```diff
@@ -13,60 +13,62 @@ import           Control.Effect.Fresh
 import           Control.Effect.NonDet
 import           Control.Effect.Reader
 import           Control.Effect.State
-import qualified Data.Core as Core
 import qualified Data.Map as Map
 import           Data.Maybe (fromMaybe)
 import           Data.Monoid (Alt(..))
 import qualified Data.Set as Set
-import           Data.Term (Term)
 
-type Cache name a = Map.Map (Term (Core.Ann :+: Core.Core) name) (Set.Set a)
-type Heap name a = Map.Map name (Set.Set a)
+newtype Cache term a = Cache { unCache :: Map.Map term (Set.Set a) }
+  deriving (Eq, Ord, Show)
+
+type Heap address a = Map.Map address (Set.Set a)
 
 newtype FrameId name = FrameId { unFrameId :: name }
   deriving (Eq, Ord, Show)
 
 
-convergeTerm :: forall m sig a name
+convergeTerm :: forall m sig a term address proxy
              .  ( Carrier sig m
                 , Effect sig
+                , Eq address
                 , Member Fresh sig
-                , Member (State (Heap name a)) sig
+                , Member (State (Heap address a)) sig
                 , Ord a
-                , Ord name
+                , Ord term
                 )
-             => (Term (Core.Ann :+: Core.Core) name -> NonDetC (ReaderC (Cache name a) (StateC (Cache name a) m)) a)
-             -> Term (Core.Ann :+: Core.Core) name
+             => proxy address
```

Comment:
We didn’t mention the address type anywhere else, so we needed a dummy parameter to keep it unambiguous. (We could also have used `AllowAmbiguousTypes` & `TypeApplications`, but I try to avoid the former since it makes calling the thing significantly more challenging.)

Reply:
+1 on avoiding `AllowAmbiguousTypes`, since `Proxy` has nice behavior with `TypeApplications` anyway.